Separate _TSObject into conversion

jbrockmendel opened the pull request:
This is what we've been building towards folks.  `convert_to_tsobject`, `convert_str_to_tsobject`, and `convert_datetime_to_tsobject` are moved to `tslibs.conversion`.  Concerns regarding `_TSObjects`s can be considered separated.

This isn't quite cut/paste:
- `convert_to_tsobject` calls `is_timestamp` which is not available in the namespace.  Similarly `_localize_pydatetime` checks `isinstance(dt, Timestamp`).  So this re-implements private versions:
- `conversion._is_timestamp` checks that its input has type `datetime`.  IIUC correctly this check should be as performant as `tslib.is_timestamp`.  (moderate sized "if")
- `conversion._localize_pydatetime` includes small optimizations that are available because we know the context in which it is being called.  Specifically, 1) it can skip the `Timestamp` case since any nanos get dropped in `pydatetime_to_dt64`, and 2) the `tz` arg is typed as a `tzinfo` object, 3) it skips unnecessary `None` check.

Other notes:
- `_get_datetime64_nanos` is called by `convert_to_tsobject`, so also had to be moved.  It fits well with the "conversion" theme.

jbrockmendel committed:
move convert_to_tsobject et al to tslibs.conversion

jbrockmendel committed:
flake8 cleanup

jbrockmendel committed:
re-declare datetimes

gfyoung commented:
+1 for refactoring and logic decoupling, but let's make sure performance didn't take a hit by accident (FYI, if you do plan on doing anymore of these types of PR's, I would suggest you provide that from the get go).

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -32,6 +50,31 @@ cdef int64_t DAY_NS = 86400000000000LL
+
+
+# TODO: We can type the input as a np.datetime64 right?
+# and the output as an int64_t?
```
yes

jreback replied:
add a doc-string

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -48,13 +91,227 @@ cdef class _TSObject:
             return self.value
 
 
+# helper to extract datetime and int64 from several different possibilities
```
so I would rather you integrate the comments with the doc-strings; in this case your comment is not useful as the doc-string is pretty good

jbrockmendel replied:
No big deal, but the comment was cut/pasted from the original.  Will change.

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -48,13 +91,227 @@ cdef class _TSObject:
+cdef _TSObject convert_str_to_tsobject(object ts, object tz, object unit,
+                                       bint dayfirst=False,
+                                       bint yearfirst=False):
+    """ ts must be a string """
```
can you update doc-string

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -32,6 +50,31 @@ cdef int64_t DAY_NS = 86400000000000LL
+    if unit != PANDAS_FR_ns:
+        pandas_datetime_to_datetimestruct(ival, unit, &dts)
+        check_dts_bounds(&dts)
+        return dtstruct_to_dt64(&dts)
```
```
    ival = dstruct_to_dt64(&dts)

return ival
```

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -32,6 +50,31 @@ cdef int64_t DAY_NS = 86400000000000LL
+# ----------------------------------------------------------------------
+# Misc Helpers
+
+cdef inline bint _is_timestamp(datetime obj):
```
rename this to make it clear

``is_instance_timestamp``

jreback replied:
if tihs is *onlyI used in that 1 place I would just in-line it directly.

jbrockmendel replied:
> if tihs is *onlyI used in that 1 place I would just in-line it directly.

Makes sense.

Just realized: recall recently there was an issue with `Timestamp` classmethods returning `Timestamp` instead of `cls` instances.  The conclusion was that this could not be fixed because `is_timestamp` would break on subclasses.  If this implementation is in fact equally performant, that would solve the problem.

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -48,13 +91,227 @@ cdef class _TSObject:
+        offset = get_utcoffset(obj.tzinfo, ts)
+        obj.value -= int(offset.total_seconds() * 1e9)
+
+    if _is_timestamp(ts):
```
comment here

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -32,6 +50,31 @@ cdef int64_t DAY_NS = 86400000000000LL
+        check_dts_bounds(&dts)
+        return dtstruct_to_dt64(&dts)
+    else:
+        return ival
```
de-privatize

note I am not only suggesting these things for general cleanliness, but also because these prevent accidental imports; IOW things will immediately break because of the new name unless they are fixed

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into tslibs-conversion10

jbrockmendel committed:
per reviewer requests:

write decent docstrings

de-private _get_datetime64_nanos

also, removed unnecessary unit input from convert_str_to_tsobject

jbrockmendel committed:
whitespace fixup

jbrockmendel committed:
fix extra arg

jbrockmendel committed:
revert not-so-bright removal of unit arg

jbrockmendel commented:
Appveyor error looks like it can't import scipy; saw the same error in #18016 a few minutes ago.

Not sure if its related, but trying to run asv it keeps timing out, stalling at the "Creating environments" step.

jbrockmendel committed:
catch ValueError

jbrockmendel commented:
Suggestions for the scipy import error on appveyor?

gfyoung commented:
@jreback : This looks like a `scipy` issue to me, as I'm seeing that in my PR as well.

jbrockmendel commented:
Forgot to use affinity on the first run here:
```
asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
      before           after         ratio
     [881ee30b]       [b129a9e1]
+     8.95±0.06μs      11.0±0.07μs     1.23  timeseries.Offsets.time_timeseries_year_apply
+       163±0.2ms        187±0.4ms     1.15  timeseries.ToDatetime.time_iso8601_tz_spaceformat
-      34.5±0.1μs      31.1±0.07μs     0.90  timeseries.Offsets.time_custom_bday_cal_decr
-      59.1±0.4μs      53.1±0.04μs     0.90  timeseries.SemiMonthOffset.time_end_incr
-           1.51s          405±9ms     0.27  timeseries.AsOfDataFrame.time_asof
-           2.00s        421±0.8ms     0.21  timeseries.AsOfDataFrame.time_asof_nan
```

Note master got updated to a new commit between runs here.

```
taskset 5 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
       before           after         ratio
     [15fa4bd6]       [b129a9e1]
+       96.1±30ms            154ms     1.61  timeseries.DatetimeIndex.time_dti_tz_factorize
+           1.71s            2.06s     1.20  timeseries.AsOfDataFrame.time_asof_nan
-           2.17s       1.95±0.03s     0.90  timeseries.AsOfDataFrame.time_asof
-      29.0±0.1μs       24.8±0.1μs     0.86  timeseries.AsOf.time_asof_single
-          51.7ms       8.82±0.4ms     0.17  timeseries.DatetimeIndex.time_dti_factorize

taskset 5 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
      before           after         ratio
     [15fa4bd6]       [b129a9e1]
+        25.5±2ms         36.0±2ms     1.41  timeseries.DatetimeIndex.time_dti_tz_factorize
-      25.6±0.2μs       22.3±0.2μs     0.87  timeseries.Offsets.time_timeseries_day_apply
-     4.08±0.01ms      3.41±0.01ms     0.83  timeseries.ToDatetime.time_iso8601
-        509±50ms          413±1ms     0.81  timeseries.DatetimeIndex.time_add_offset_slow

taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries
[...]
       before           after         ratio
     [15fa4bd6]       [b129a9e1]
+      11.9±0.2μs       14.8±0.3μs     1.24  timeseries.Offsets.time_timeseries_year_incr
+       173±0.4ms          198±5ms     1.15  timeseries.ToDatetime.time_iso8601_tz_spaceformat
+     7.15±0.05ms       8.04±0.4ms     1.12  timeseries.DatetimeIndex.time_add_offset_fast
+      49.0±0.4μs         54.3±1μs     1.11  timeseries.SemiMonthOffset.time_begin_apply
-        539±20μs        477±0.4μs     0.89  timeseries.DatetimeIndex.time_reset_index_tz
-      444±0.01μs        387±0.3μs     0.87  timeseries.DatetimeIndex.time_reset_index
-      22.8±0.2μs      19.7±0.09μs     0.87  timeseries.Offsets.time_custom_bday_apply_dt64
-      2.21±0.2ms         1.72±0ms     0.78  timeseries.DatetimeIndex.time_add_offset_delta
```

Given the excess variability caused by my local entropy field, this looks about unchanged.

jreback commented:
looks good. pls rebase

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into tslibs-conversion10

jreback commented:
thanks!
